assembly_uploader/study_xmls.py

```diff
@@ -45,48 +53,79 @@ def parse_args(argv):
     parser.add_argument(
         "--publication",
         help="pubmed ID for connected publication if available",
+        type=int,
         required=False,
     )
     parser.add_argument("--output-dir", help="Path to output directory", required=False)
     return parser.parse_args(argv)
 
 
-class RegisterStudy:
-    def __init__(self, argv=sys.argv[1:]):
-        self.args = parse_args(argv)
-        self.study = self.args.study
-        if self.args.output_dir:
-            self.upload_dir = os.path.join(self.args.output_dir, f"{self.study}_upload")
-        else:
-            self.upload_dir = os.path.join(os.getcwd(), f"{self.study}_upload")
-        self.study_xml_path = os.path.join(self.upload_dir, f"{self.study}_reg.xml")
-        self.submission_xml_path = os.path.join(
-            self.upload_dir, f"{self.study}_submission.xml"
+class StudyXMLGenerator:
+    def __init__(
+        self,
+        study: str,
+        center_name: str,
+        library: str,
+        hold_date: datetime = None,
+        tpa: bool = False,
+        output_dir: Path = None,
+        publication: int = None,
+    ):
+        f"""
+        Build submission files for an assembly study.
+
+        :param study: raw reads study ID/accession
+        :param center_name: submission centre name, e.g. EMG
+        :param library: {METAGENOME} or {METATRANSCRIPTOME}
+        :param hold_date: hold date for the data to remain private, if it should be different from the provided study"
+        :param tpa: is this a third-party assembly?
+        :param output_dir: path to output directory (default is CWD)
+        :param publication: pubmed ID for connected publication if available
+        :return: StudyXMLGenerator object
+        """
+        self.study = study
+
+        self.upload_dir = (output_dir or Path(".")) / Path(f"{self.study}_upload")
+        self.upload_dir = self.upload_dir.absolute()
+        self.upload_dir.mkdir(parents=True, exist_ok=True)
+
+        self.study_xml_path = self.upload_dir / Path(f"{self.study}_reg.xml")
+        self.submission_xml_path = self.upload_dir / Path(
+            f"{self.study}_submission.xml"
         )
-        self.center = self.args.center
-        self.hold = self.args.hold
+
+        self.center = center_name
+        self.hold_date = hold_date
+
+        assert library in [METAGENOME, METATRANSCRIPTOME]
```

Comment:
Good q – sort of. I think it makes sense here because: in CLI mode, the user input of `library` is already validated because I added `choices=` to the `argparse`. (Generally you should not use `assert`ions to check user input, because assertion checks can be disabled at runtime by running python in "optimise" mode.) So here the assertion can only fail through programming error – where assertions are useful to sanity check things that should always be true. Would be slightly better if the "library user" input was validated differently (e.g. a Pydantic type or explicit check) but in my opinion this kind of check is okay for that, and the stacktrace for this failing should be self explanatory.